### PR TITLE
WIP: Switch to dynamic plugins

### LINKED ISSUE #528: Switch  static plugins in the showcase backend  to dynamic plugins
as much plugins as possible to switch in a sprint

the goal is to open PR (wip) with all the backend plugins switched and docker image available for testing.

### PR DESCRIPTION
**Pull request by davidfestal**
Signed-off-by: David Festal <dfestal@redhat.com>

## Description

Switch static plugins to dynamic plugins, enabled by default.

## Which issue(s) does this PR fix

- Fixes #528

## PR acceptance criteria

Please make sure that the following steps are complete:

- [ ] GitHub Actions are completed and successful
- [ ] Unit Tests are updated and passing
- [ ] E2E Tests are updated and passing
- [ ] Documentation is updated if necessary (requirement for new features)
- [ ] Add a screenshot if the change is UX/UI related

### CODE REVIEW

**Comment by nickboldt** [comment]
Downstream: [RHIDP-429](https://issues.redhat.com/browse/RHIDP-429) Adapt RHDH build for building dynamic plugins

File: docker/install-dynamic-plugins.py [hunk 1]
```diff
@@ -9,16 +9,16 @@ class InstallException(Exception):
     """Exception class from which every exception in this library will derive."""
     pass
 
-def merge(source, destination):
+def merge(source, destination, prefix = ''):
```
Comment (nickboldt) [inline docker/install-dynamic-plugins.py:12]: Can you add a standard copyright header and small usage explanation for this script so others know what it's for, how it's used, etc.  ?

File: docker/Dockerfile [hunk 2]
```diff
@@ -46,6 +46,8 @@ RUN chmod +x $YARN
 # Stage 2 - Install dependencies
 FROM skeleton AS deps
 
+COPY $EXTERNAL_SOURCE_NESTED/dynamic-plugins/ ./dynamic-plugins/
+#RUN find dynamic-plugins \! -name "package.json" -mindepth 2 -maxdepth 2 -print | xargs rm -rf
```
Comment (nickboldt) [inline docker/Dockerfile:50]: if this is a useful cleanup step, could move this down to line 102 after

```
RUN $YARN install --frozen-lockfile --production --network-timeout 600000
```

Comment (davidfestal) [reply docker/Dockerfile:50]: I think it's not useful in fact.

**Comment by Zaperex** [comment]
<details><summary>Sometimes, whenever I restart rollouts on a deployment, the initial pod always fails with the following error logs:</summary>

```console
2023-10-13T14:48:21.609Z backstage error error while watching '/opt/app-root/src/dynamic-plugins-root' ENOSPC: System limit for number of file watchers reached, watch '/opt/app-root/src/dynamic-plugins-root/plugin-techdocs-backend-wrapped-dynamic-1.7.1-dynamic.0/node_modules/lodash/zip.js' errno=-28 syscall=watch code=ENOSPC path=/opt/app-root/src/dynamic-plugins-root/plugin-techdocs-backend-wrapped-dynamic-1.7.1-dynamic.0/node_modules/lodash/zip.js filename=/opt/app-root/src/dynamic-plugins-root/plugin-techdocs-backend-wrapped-dynamic-1.7.1-dynamic.0/node_modules/lodash/zip.js stack=Error: ENOSPC: System limit for number of file watchers reached, watch '/opt/app-root/src/dynamic-plugins-root/plugin-techdocs-backend-wrapped-dynamic-1.7.1-dynamic.0/node_modules/lodash/zip.js'
    at FSWatcher.<computed> (node:internal/fs/watchers:247:19)
    at Object.watch (node:fs:2418:34)
    at createFsWatchInstance (/opt/app-root/src/node_modules/chokidar/lib/nodefs-handler.js:119:15)
    at setFsWatchListener (/opt/app-root/src/node_modules/chokidar/lib/nodefs-handler.js:166:15)
    at NodeFsHandler._watchWithNodeFs (/opt/app-root/src/node_modules/chokidar/lib/nodefs-handler.js:331:14)
    at NodeFsHandler._handleFile (/opt/app-root/src/node_modules/chokidar/lib/nodefs-handler.js:395:23)
    at NodeFsHandler._addToNodeFs (/opt/app-root/src/node_modules/chokidar/lib/nodefs-handler.js:637:21)
node:internal/process/promises:288
            triggerUncaughtException(err, true /* fromPromise */);
            ^

Error: ENOSPC: System limit for number of file watchers reached, watch '/opt/app-root/src/dynamic-plugins-root/plugin-techdocs-backend-wrapped-dynamic-1.7.1-dynamic.0/node_modules/lodash/zip.js'
    at FSWatcher.<computed> (node:internal/fs/watchers:247:19)
    at Object.watch (node:fs:2418:34)
    at createFsWatchInstance (/opt/app-root/src/node_modules/chokidar/lib/nodefs-handler.js:119:15)
    at setFsWatchListener (/opt/app-root/src/node_modules/chokidar/lib/nodefs-handler.js:166:15)
    at NodeFsHandler._watchWithNodeFs (/opt/app-root/src/node_modules/chokidar/lib/nodefs-handler.js:331:14)
    at NodeFsHandler._handleFile (/opt/app-root/src/node_modules/chokidar/lib/nodefs-handler.js:395:23)
    at NodeFsHandler._addToNodeFs (/opt/app-root/src/node_modules/chokidar/lib/nodefs-handler.js:637:21) {
  errno: -28,
  syscall: 'watch',
  code: 'ENOSPC',
  path: '/opt/app-root/src/dynamic-plugins-root/plugin-techdocs-backend-wrapped-dynamic-1.7.1-dynamic.0/node_modules/lodash/zip.js',
  filename: '/opt/app-root/src/dynamic-plugins-root/plugin-techdocs-backend-wrapped-dynamic-1.7.1-dynamic.0/node_modules/lodash/zip.js'
}

Node.js v18.17.1
```
</details>

However, outside of the initial pod of each deployment, this error doesn't occur again.

**Comment by davidfestal** [comment]
Closed since it was replaced by https://github.com/janus-idp/backstage-showcase/pull/631 which has been merged.

**Pull request closed without merging**